lib/timers.js

```diff
@@ -226,122 +284,105 @@ function listOnTimeout(handle, now) {
-      return true;
+      return;
     }
 
     // The actual logic for when a timeout happens.
-
     L.remove(timer);
-    assert(timer !== L.peek(list));
+
+    const asyncId = timer[async_id_symbol];
 
     if (!timer._onTimeout) {
-      if (destroyHooksExist() && !timer._destroyed &&
-            typeof timer[async_id_symbol] === 'number') {
-        emitDestroy(timer[async_id_symbol]);
+      if (timer[kRefed])
+        refCount--;
+      timer[kRefed] = null;
+
+      if (destroyHooksExist() && !timer._destroyed) {
+        emitDestroy(asyncId);
         timer._destroyed = true;
       }
       continue;
     }
 
+    emitBefore(asyncId, timer[trigger_async_id_symbol]);
+
     tryOnTimeout(timer);
+
+    emitAfter(asyncId);
   }
 
   // If `L.peek(list)` returned nothing, the list was either empty or we have
   // called all of the timer timeouts.
-  // As such, we can remove the list and clean up the TimerWrap C++ handle.
+  // As such, we can remove the list from the object map and the PriorityQueue.
   debug('%d list empty', msecs);
-  assert(L.isEmpty(list));
-
-  // Either refedLists[msecs] or unrefedLists[msecs] may have been removed and
-  // recreated since the reference to `list` was created. Make sure they're
-  // the same instance of the list before destroying.
-  if (list._unrefed === true && list === unrefedLists[msecs]) {
-    delete unrefedLists[msecs];
-  } else if (list === refedLists[msecs]) {
-    delete refedLists[msecs];
-  }
-
-  // Do not close the underlying handle if its ownership has changed
-  // (e.g it was unrefed in its callback).
-  if (!handle.owner)
-    handle.close();
 
-  return true;
+  // The current list may have been removed and recreated since the reference
+  // to `list` was created. Make sure they're the same instance of the list
+  // before destroying.
+  if (list === lists[msecs]) {
+    delete lists[msecs];
+    queue.shift();
+  }
 }
 
 
 // An optimization so that the try/finally only de-optimizes (since at least v8
 // 4.7) what is in this smaller function.
 function tryOnTimeout(timer, start) {
-  timer._called = true;
-  const timerAsyncId = (typeof timer[async_id_symbol] === 'number') ?
-    timer[async_id_symbol] : null;
-  var threw = true;
-  if (timerAsyncId !== null)
-    emitBefore(timerAsyncId, timer[trigger_async_id_symbol]);
   if (start === undefined && timer._repeat)
     start = TimerWrap.now();
   try {
     ontimeout(timer);
-    threw = false;
   } finally {
-    if (timerAsyncId !== null) {
-      if (!threw)
-        emitAfter(timerAsyncId);
-      if (timer._repeat) {
-        rearm(timer, start);
-      } else if (destroyHooksExist() && !timer._destroyed) {
-        emitDestroy(timerAsyncId);
+    if (timer._repeat) {
+      rearm(timer, start);
+    } else {
+      if (timer[kRefed])
+        refCount--;
+      timer[kRefed] = null;
+
+      if (destroyHooksExist() && !timer._destroyed) {
+        emitDestroy(timer[async_id_symbol]);
```

Comment:
This is wrong, since the `after` callback from `listOnTimeout` will be emitted later than the `destroy` from here.

Reply:
Destroy is emitted in an async manner. (See next tick for similar scenario.)

Reply:
Ok... but it seems weird that the testcases for this would then have changed.

Reply:
Only the TimerWrap test cases have changed as far as async hooks go. The ones for Timeouts remain the same.